source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
   return Network::PostIoAction::KeepOpen;
 }
 
+Network::IoResult TsiSocket::repeatReadAndUnprotect(Buffer::Instance& buffer,
+                                                    Network::IoResult prev_result) {
+  Network::IoResult result = prev_result;
+  uint64_t total_bytes_processed = 0;
+
+  do {
+    // Do unprotect.
+    uint64_t prev_size = buffer.length();
+    ENVOY_CONN_LOG(debug, "TSI: unprotecting buffer size: {}", callbacks_->connection(),
+                   raw_read_buffer_.length());
+    tsi_result status = frame_protector_->unprotect(raw_read_buffer_, buffer);
+    if (status != TSI_OK) {
+      break;
+    }
+    ENVOY_CONN_LOG(debug, "TSI: unprotected buffer left: {} result: {}", callbacks_->connection(),
+                   raw_read_buffer_.length(), tsi_result_to_string(status));
+    total_bytes_processed += buffer.length() - prev_size;
+    if (result.action_ == Network::PostIoAction::Close) {
+      break;
+    }
+
+    // Check if buffer needs to be drained.
+    if (callbacks_->shouldDrainReadBuffer()) {
+      callbacks_->setTransportSocketIsReadable();
+      break;
+    } else {
+      // Do another read.
+      if (!end_stream_read_ && !read_error_) {
+        result = raw_buffer_socket_->doRead(raw_read_buffer_);
+        end_stream_read_ = result.end_stream_read_;
+        read_error_ = result.action_ == Network::PostIoAction::Close;
+        // No data is read.
+        if (result.bytes_processed_ == 0) {
+          break;
+        }
+        // Read error happens or end-of-stream is reached in the previous read.
+      } else {
+        break;
+      }
+    }
+  } while (true);
+
+  result.bytes_processed_ = total_bytes_processed;
+  return result;
+}
+
 Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
   Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
-  if (!end_stream_read_ && !read_error_) {
+  if (!handshake_complete_ && !end_stream_read_ && !read_error_) {
```

Comment:
Agree on redundant !handshake_complete_ check. But I do not think we should remove other checks for !handshake_complete_ out of the body of "if (!handshake_complete_)" because they can be used to return early without doing handshake.